ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,440 @@
+
+        self.num_detection_max = num_detection_max
+
+        if iou_thresholds is None:
+            iou_thresholds = [0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95]
+        self.iou_thresholds = torch.tensor(iou_thresholds, device=device)
+
+        if rec_thresholds is None:
+            rec_thresholds = [i / 100 for i in range(101)]
+        self.rec_thresholds = torch.tensor(rec_thresholds, device=device)
+
+        super(MeanAveragePrecision, self).__init__(output_transform=output_transform, device=device)
+
+    def _check_object_area_ranges(self) -> None:
+        for area, area_range in self.object_area_ranges.items():
+            if len(area_range) != 2 or area_range[0] >= area_range[1]:
+                raise ValueError(
+                    f"object_area_ranges must be a dict associating to each key (str) a tuple of float values (a, b) "
+                    f"where a < b (got: key={area}, value={area_range})"
+                )
+
+    @reinit__is_reduced
+    def update(self, outputs: Tuple[Dict, Dict]) -> None:
+
+        for output in outputs:
+
+            y_pred_img, y_img = output
+
+            assert y_img["image_id"] == y_pred_img["image_id"]
+
+            if y_img["image_id"] in self.image_ids:
+                raise ValueError("Detections for this image_id are already evaluated.")
+
+            self.image_ids.add(y_img["image_id"])
+
+            y_category_dict = defaultdict(list)
+            for i, category_id in enumerate(y_img["category_id"]):
+                y_category_dict[int(category_id)].append(i)
+
+            y_pred_category_dict = defaultdict(list)
+            for i, category_id in enumerate(y_pred_img["category_id"]):
+                y_pred_category_dict[int(category_id)].append(i)
+
+            categories = torch.unique(torch.cat([y_img["category_id"], y_pred_img["category_id"]])).tolist()
+            self.category_ids.update(categories)
+
+            for category in categories:
+                y_pred_ind = y_pred_category_dict[category]
+                y_pred_bbox = y_pred_img["bbox"][y_pred_ind]
+                y_pred_score = y_pred_img["score"][y_pred_ind]
+                y_pred_id = y_pred_img["id"][y_pred_ind]
+                y_pred_area = (y_pred_img["bbox"][:, 2] * y_pred_img["bbox"][:, 3])[y_pred_ind]
+
+                sorted_y_pred_bbox = y_pred_bbox[torch.argsort(-y_pred_img["score"][y_pred_ind])]
+
+                y_ind = y_category_dict[category]
+                y_id = y_img["id"][y_ind]
+                y_bbox = y_img["bbox"][y_ind]
+                crowd = y_img["iscrowd"][y_ind] if "iscrowd" in y_img else None
+                y_ignore = y_img["ignore"][y_ind] if "ignore" in y_img else torch.zeros(len(y_ind))
+                y_area = y_img["area"][y_ind] if "area" in y_img else (y_img["bbox"][:, 2] * y_img["bbox"][:, 3])[y_ind]
+
+                ious = _iou(y_bbox, sorted_y_pred_bbox, crowd).to(self._device)
```

Comment:
I don’t see why y_pred’s boxes should be sorted by score. Moreover, that is not mentioned in the iou method.

Reply:
iou method is generic, for 'a' y_bbox and 'b' y_pred_bbox it return a matrix of size 'a x b' containing all the respective ious. We are passing it sorted list because in eval_image function we are sorting y_preds according to confidence as well for greedy matching.

Reply:
I don’t see where it matters…